### ISSUE TITLE
How to build from source

### ISSUE THREAD

**Issue by bpali** [traccar/traccar#2802]
I'm running a CentOS 7 64bit machine and I've read the [documentation](https://www.traccar.org/build/) about how to build from source.

Still I'm opening this issue because the documentation seems to be incomplete or maybe I'm not understanding it the right way.

I have a local repository clone of the source files at the latest version. I make no modifications to the source code here so I can always pull the master branch to its head revision.

I run "mvn package" on it and see the output in the target dir.

My dilemma starts here: Is it enough to overwrite tracker-server.jar from the target dir to my /opt/traccar running instance? (I'm doing this off course after the traccar service is stopped) do I have to take do something else?

Do I have to put anything in the /opt/traccar/lib folder?

I did the above in the past and it worked, now it doesn't.

I know I'm not doing something and I guess the success due to just overwriting tracker-server.jar depends on how far the master branch is from the latest release and if there are any major things that changed (like a new dependency created or something added to the lib folder)

I forgot to mention that I always overwrite the latest stable release version

In short words my issue is about how to build and run the head revision of the master branch

Thanks and appreciate any answer

**Comment by tananaev** [comment]
If you are building from the latest source code, you should take all the files (`lib`, `schema`, `templates` etc etc).

**Comment by hackerunet** [comment]
isn't just enough with git pull ??

**Comment by tananaev** [comment]
@hackerunet, are you sure you read the question properly?

**Comment by hackerunet** [comment]
I'm sorry, yes I did, but not the answer, ok, please consider this:
1) Permissions over the folder and subfolders, are you really overwritting the target tracker-server.jar file?
2) after building from source compare both files diff FILE1 FILE2 
3) trying removing the /opt/traccar/tracker-server.jar file with rm -rf and after that mv tracker-server.jar /opt/traccar/
That's my only suggestion about how to make sure you are replacing the file correctly.

**Comment by hackerunet** [comment]
Also one final advice, just stop using CentOS and start using debian, I used to work with centos for over 4 years and I was forced to move from it to debian, centos with all it's security enhancements and SE linux sometimes are very hard to control if you don't have enough knowledge about it. is some times very frustrating and also very limited, the top allowed files descriptors aren't good enough for traccar and stuff starts failing with time.
best wishes.

**Comment by bpali** [comment]
Thanks @tananaev and @hackerunet for your answers but to clarify a bit things here let's consider this scenario:
- I have a linux server
- I have a cloned repository of https://github.com/tananaev/traccar on that server
- the repository is on branch master and is updated to the latest commit

Now: How do I build a traccar server from source code?
1. I run **mvn package** on the respository
1. mvn runs OK and I have the results in the **target** folder
1.  ... what do I do next ???

**Comment by tananaev** [comment]
Now you can take JAR file and lib folder and replace it on your installation.

**Comment by seym45** [comment]
@bpali  what did you do next? any luck?
I have done same as you, now what?

**Comment by bpali** [comment]
@seym45 I've replaced the JAR file and lib folder into my installation and it worked for what I wanted to achieve. What is you use case, explain better what you actually want to do and I will try to guide you.

**Comment by SebastEnn** [comment]
He probably figured it out. Issue can be closed.
